In [1]:
import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [2]:
# loading data
df = pd.read_csv('/kaggle/input/international-football-results-from-1872-to-2017/results.csv')
print(df.shape)
df.head()

(44762, 9)


date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

In [3]:
# check for null values
df.isna().sum()

date          0
home_team     0
away_team     0
home_score    0
away_score    0
tournament    0
city          0
country       0
neutral       0
dtype: int64

In [4]:
# remove null values
df.dropna(inplace=True)
print(df.isna().sum().sum())

0


In [5]:
# basic info about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44762 entries, 0 to 44761
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        44762 non-null  object
 1   home_team   44762 non-null  object
 2   away_team   44762 non-null  object
 3   home_score  44762 non-null  int64 
 4   away_score  44762 non-null  int64 
 5   tournament  44762 non-null  object
 6   city        44762 non-null  object
 7   country     44762 non-null  object
 8   neutral     44762 non-null  bool  
dtypes: bool(1), int64(2), object(6)
memory usage: 2.8+ MB


In [6]:
# teams that played the number of away matches
df.away_team.value_counts()

Uruguay          562
Sweden           548
England          526
Hungary          492
Paraguay         475
                ... 
Andalusia          1
Asturias           1
Central Spain      1
Corsica            1
Aymara             1
Name: away_team, Length: 306, dtype: int64

In [7]:
tournaments = df.tournament.unique().tolist()
print((f'There are {len(tournaments)} tournaments in this dataset'))

There are 145 tournaments in this dataset


In [8]:
# most games are played in these cities 
df.city.value_counts()[:10]

Kuala Lumpur    677
Doha            519
Bangkok         429
London          420
Budapest        414
Montevideo      363
Kuwait City     356
Vienna          353
Glasgow         342
Copenhagen      342
Name: city, dtype: int64

In [9]:
# parsing dates
df['date'] = pd.to_datetime(df['date'],errors='coerce',format='%Y-%m-%d')

df['Year'] = df['date'].dt.year
df['month'] = df['date'].dt.month
df['day'] = df['date'].dt.day

day_order = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
df['weekday'] = df['date'].map(lambda x: day_order[x.weekday()] )

print(df.dtypes)
df.head()

date          datetime64[ns]
home_team             object
away_team             object
home_score             int64
away_score             int64
tournament            object
city                  object
country               object
neutral                 bool
Year                   int64
month                  int64
day                    int64
weekday               object
dtype: object


date home_team away_team  home_score  away_score tournament     city  \
0 1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1 1873-03-08   England  Scotland           4           2   Friendly   London   
2 1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3 1875-03-06   England  Scotland           2           2   Friendly   London   
4 1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  Year  month  day weekday  
0  Scotland    False  1872     11   30     Sat  
1   England    False  1873      3    8     Sat  
2  Scotland    False  1874      3    7     Sat  
3   England    False  1875      3    6     Sat  
4  Scotland    False  1876      3    4     Sat

In [10]:
# making a new column to denote a 'win' ,'lose' or 'tie'
def find_result(df):
    if df.home_score > df.away_score:
        return 'win'
    if df.away_score > df.home_score:
        return 'lost'
    else:
        return 'tie'
    
df['result'] = df.apply(find_result,axis=1)
df

date      home_team away_team  home_score  away_score  \
0     1872-11-30       Scotland   England           0           0   
1     1873-03-08        England  Scotland           4           2   
2     1874-03-07       Scotland   England           2           1   
3     1875-03-06        England  Scotland           2           2   
4     1876-03-04       Scotland   England           3           0   
...          ...            ...       ...         ...         ...   
44757 2023-07-09  United States    Canada           2           2   
44758 2023-07-12  United States    Panama           1           1   
44759 2023-07-12        Jamaica    Mexico           0           3   
44760 2023-07-16         Kernow     Sápmi           2           1   
44761 2023-07-16         Mexico    Panama           1           0   

                                    tournament        city        country  \
0                                     Friendly     Glasgow       Scotland   
1                                     Friendly      London        England   
2                                     Friendly     Glasgow       Scotland   
3                                     Friendly      London        England   
4                                     Friendly     Glasgow       Scotland   
...                                        ...         ...            ...   
44757                                 Gold Cup  Cincinnati  United States   
44758                                 Gold Cup   San Diego  United States   
44759                                 Gold Cup    Paradise  United States   
44760  CONIFA World Football Cup qualification    Falmouth        England   
44761                                 Gold Cup   Inglewood  United States   

       neutral  Year  month  day weekday result  
0        False  1872     11   30     Sat    tie  
1        False  1873      3    8     Sat    win  
2        False  1874      3    7     Sat    win  
3        False  1875      3    6     Sat    tie  
4        False  1876      3    4     Sat    win  
...        ...   ...    ...  ...     ...    ...  
44757    False  2023      7    9     Sun    tie  
44758    False  2023      7   12     Wed    tie  
44759     True  2023      7   12     Wed   lost  
44760    False  2023      7   16     Sun    win  
44761     True  2023      7   16     Sun    win  

[44762 rows x 14 columns]

# EDA

### 1) In which month most matches are played ?

In [11]:
fig = px.histogram(df,
                   x="month",
                   text_auto=True,
                   title='Number of matches played per month',
                   category_orders= {"month": list(range(1,13))})
fig.update_layout(bargap=0.2)
fig.update_xaxes(type='category',title_text='Number of months')
fig.update_yaxes(title_text='Matches played')
fig.show()

On June , august, september most matches are played

### 2) On which weekdays most matches are played ?

In [12]:
fig = px.histogram(df,
                   x="weekday",
                   text_auto=True,
                   title='Number of matches played per weekday',
                   category_orders= {"weekday": day_order}
                   )
fig.update_layout(bargap=0.2)
fig.update_xaxes(type='category',title_text='Number of months')
fig.update_yaxes(title_text='matches played')
fig.show()

On wednesday , sunday most matches are played

### 3) Does home advantage helps a team to win ?

In [13]:
fig = go.Figure()
fig = px.histogram(df,
           x = 'neutral',
           color='result',
           title='Venue distribution on basis of result',
           barmode='group',
           text_auto=True,
           )
fig.show()

From the above plots if we only see the win count in **false**(not a neutral venue or home ground) and **true**(neutral ground) then there is home advantage as in **false**(***win is around 17K***) and in **true**(***win = 4807***) 

### 4) Number of matches played in tournaments each year ?

In [14]:
tournaments_count_over_years = pd.DataFrame(df.groupby(['Year','tournament'])['tournament'].count()).rename(columns={'tournament':'matches_played'})
tournaments_count_over_years = tournaments_count_over_years.reset_index().sort_values('Year',ascending=False)

fig = go.Figure()
fig = px.line(tournaments_count_over_years,
              x="Year", y=["matches_played"],
              color='tournament',
              labels={'matches_played':'number of matches'})

fig.update_layout(title='Number of matches played in each tournament over the years',
                  yaxis_title='Matches played')

fig.show()

There are **609** ***FIFA World Cup qualification*** matches played in 2021 highest till now. It is followed by 2001 ***491 FIFA World Cup qualification***  matches played and then **426** ***Friendly*** matches played in 2018.

### 5) Who is the best team of all time ?

There home_teams and away_team available . So to find the best team let's find the best home_teams and away_teams then compare both.

In [15]:
best_home_team = df[df.home_score > df.away_score]
best_home_team = pd.DataFrame(best_home_team.groupby(['home_team'])['home_team'].agg('count'))
best_home_team = best_home_team.rename(columns={'home_team':'home_matches_won'}).sort_values('home_matches_won',ascending=False).reset_index()
best_home_team[:5]

home_team  home_matches_won
0     Brazil               428
1  Argentina               383
2     Mexico               330
3    Germany               328
4    England               328

In [16]:
best_away_team = df[df.home_score < df.away_score]
best_away_team = pd.DataFrame(best_away_team.groupby(['away_team'])['away_team'].agg('count'))
best_away_team = best_away_team.rename(columns={'away_team':'away_matches_won'}).sort_values('away_matches_won',ascending=False).reset_index()
best_away_team[:5]

away_team  away_matches_won
0   England               273
1   Germany               247
2    Brazil               229
3    Sweden               222
4   Uruguay               202

By comapring the top 5 home teams and away teams we can say that Brazil has won most number of home matches. As we seen before there is always home advantage as a home team playes well in it's home ground. 

### 6) Which teams dominated different eras of football ?

In [17]:
# making a new column just for the purpose of analysis
def find_era(df):
    eighteens = np.arange(1872,1900).tolist()
    nineteens = np.arange(1900,2000).tolist()
    twenties = np.arange(2000,3000).tolist()
    
    if df.Year in eighteens:
        return '18s'
    if df.Year in nineteens:
        return '19s'
    if df.Year in twenties:
        return '20s'

df['era'] = df.apply(find_era,axis=1)
df.sample(5)

date   home_team     away_team  home_score  away_score  \
25011 2002-03-10    Eswatini       Lesotho           1           0   
18944 1994-11-16  Azerbaijan        Israel           0           2   
31171 2008-10-15     Croatia       Andorra           4           0   
21999 1998-10-19        Oman  Burkina Faso           4           2   
3905  1956-01-29   Argentina         Chile           2           0   

                         tournament        city   country  neutral  Year  \
25011                      Friendly     Mbabane  Eswatini    False  2002   
18944       UEFA Euro qualification     Trabzon    Turkey     True  1994   
31171  FIFA World Cup qualification      Zagreb   Croatia    False  2008   
21999                      Friendly      Muscat      Oman    False  1998   
3905                   Copa América  Montevideo   Uruguay     True  1956   

       month  day weekday result  era  
25011      3   10     Sun    win  20s  
18944     11   16     Wed   lost  19s  
31171     10   15     Wed    win  20s  
21999     10   19     Mon    win  19s  
3905       1   29     Sun    win  19s

In [18]:
# Most TROPHIES won in each tournament by each team
# we will focus on best home teams

best_teams = df[(df.result=='win')]
best_teams = best_teams.groupby(['era','home_team'])[['home_score']].count().sort_values('home_score',ascending=False).rename(columns={'home_score':'matches_won'}).reset_index()
best_teams[:20]

era       home_team  matches_won
0   19s          Brazil          298
1   19s       Argentina          271
2   19s          Sweden          218
3   19s         Germany          217
4   19s         Hungary          213
5   19s         England          200
6   19s           Italy          197
7   19s     South Korea          179
8   19s          France          172
9   20s          Mexico          165
10  19s          Mexico          165
11  20s   United States          165
12  19s         Austria          164
13  19s         Denmark          160
14  19s     Netherlands          152
15  19s        Scotland          147
16  19s         Belgium          147
17  20s           Egypt          143
18  19s  Czechoslovakia          140
19  19s         Uruguay          139

In [19]:
fig = px.bar(best_teams, x="home_team",
             y = 'matches_won' ,
             color='era',
             title='Total matches won by teams in each era'
             )
fig.show()

We can see that in the **19s** era(or from 1900-2000) **Brazil** grabbed the top spot followed by ***Argentina , Sweden , Germany***. Many legends played in that era like from **Brazil** we had ***Pele,Garrincha,Ronaldo*** etc. while from **Argentina** we had ***Diego Maradona,Alfredo di Stefano,Gabriel Batistuta***,etc. Also advancements in televisions can be a reason for increasing footbal's popularity. 

I want to find which are the best teams in world's top world tournaments

In [20]:
top_tours = df[
   (df.tournament == 'FIFA World Cup') |
   (df.tournament == 'UEFA Nations League')|
   (df.tournament == 'Copa América')|
   (df.tournament == 'UEFA Euro')|
   (df.tournament == 'African Cup of Nations')|
   (df.tournament == 'Confederations Cup')|
   (df.tournament == "King's Cup")|
   (df.tournament == 'AFC Asian Cup')
  ]

top_tours

date    home_team away_team  home_score  away_score  \
438   1916-07-02        Chile   Uruguay           0           4   
440   1916-07-06    Argentina     Chile           6           1   
441   1916-07-08       Brazil     Chile           1           1   
442   1916-07-10    Argentina    Brazil           1           1   
444   1916-07-12       Brazil   Uruguay           1           2   
...          ...          ...       ...         ...         ...   
44358 2022-12-18    Argentina    France           3           3   
44596 2023-06-14  Netherlands   Croatia           2           4   
44611 2023-06-15        Spain     Italy           2           1   
44663 2023-06-18  Netherlands     Italy           2           3   
44664 2023-06-18      Croatia     Spain           0           0   

                tournament          city      country  neutral  Year  month  \
438           Copa América  Buenos Aires    Argentina     True  1916      7   
440           Copa América  Buenos Aires    Argentina    False  1916      7   
441           Copa América  Buenos Aires    Argentina     True  1916      7   
442           Copa América  Buenos Aires    Argentina    False  1916      7   
444           Copa América  Buenos Aires    Argentina     True  1916      7   
...                    ...           ...          ...      ...   ...    ...   
44358       FIFA World Cup        Lusail        Qatar     True  2022     12   
44596  UEFA Nations League     Rotterdam  Netherlands    False  2023      6   
44611  UEFA Nations League      Enschede  Netherlands     True  2023      6   
44663  UEFA Nations League      Enschede  Netherlands    False  2023      6   
44664  UEFA Nations League     Rotterdam  Netherlands     True  2023      6   

       day weekday result  era  
438      2     Sun   lost  19s  
440      6     Thu    win  19s  
441      8     Sat    tie  19s  
442     10     Mon    tie  19s  
444     12     Wed   lost  19s  
...    ...     ...    ...  ...  
44358   18     Sun    tie  20s  
44596   14     Wed   lost  20s  
44611   15     Thu    win  20s  
44663   18     Sun   lost  20s  
44664   18     Sun    tie  20s  

[4104 rows x 15 columns]

In [21]:
winning_home_teams = top_tours[(top_tours.result == 'win')].groupby(['tournament','home_team'])[['home_team']].count().rename(columns={'home_team':'total_matches_won'})

rows = 8
fig = make_subplots(
    rows=rows, cols=1,
    subplot_titles=("AFC Asian Cup", "African Cup of Nations", "Confederations Cup", "Copa América",'FIFA World Cup',"King's Cup",'UEFA Nations League','UEFA Euro'))


cups = ["AFC Asian Cup", "African Cup of Nations", "Confederations Cup", "Copa América",'FIFA World Cup',"King's Cup",'UEFA Nations League','UEFA Euro']

# plot the top 10 teams to won most number of matches in a tournament
for idx,cup in enumerate(cups):
    fig.add_trace(go.Bar(
                     x=winning_home_teams.loc[cup].sort_values('total_matches_won',ascending=False)[:10].index, 
                     y=winning_home_teams.loc[cup].sort_values('total_matches_won',ascending=False)[:10].total_matches_won,
                     name=cup,
                     showlegend=False,),
                     row=idx+1, col=1)

fig.update_layout(height=1100,
                  width=1300,
                  title_text="Top 10 Home Teams winning most matches in particular tournament")

fig.show()

### 7) What countries play with each other the most ?

In [22]:
most_matches_together = df.groupby(['home_team','away_team'])[['home_team']].agg('count').rename(columns={'home_team' : 'matches_played_together'}).sort_values('matches_played_together',ascending=False).reset_index()
most_matches_together[:10]

home_team    away_team  matches_played_together
0    Argentina      Uruguay                       98
1      Uruguay    Argentina                       81
2      Austria      Hungary                       69
3      Hungary      Austria                       68
4        Kenya       Uganda                       66
5      Belgium  Netherlands                       64
6    Argentina     Paraguay                       64
7  Netherlands      Belgium                       63
8       Brazil     Paraguay                       60
9      England     Scotland                       59

**Argentina** and **Uruguay** have grabbed the 1st spot with ***179*** total matches together followed by **Austria** and **Hungry** with ***137*** matches together.

### 8) Which countries host the most matches where they themselves are not participating in ?

In [23]:
not_participated_matches = df[(df.neutral==True)]['country'].value_counts().reset_index().rename(columns={'index':'country',
                                                                               'country':'matches_hosted'})

fig = px.bar(not_participated_matches[:50],
           x = 'country',
           y = 'matches_hosted',
           color='country',
           title='Top 50 countries that hosted matches in which it not participated',
           text_auto=True,
           )
fig.update_traces(textfont_size=12, textangle=0, textposition="outside", cliponaxis=False)

fig.show()

**US** has hosted 894 matches in which it not participated. It is followed by **Malaysia**(473 matches) and **Quatar**(372 matches). 

### 9) Which teams are the most active in playing friendlies and friendly tournaments ?

In [24]:
friendly_matches = df[df.tournament == 'Friendly'].groupby(['home_team','away_team'])[['home_team']].agg('count').rename(columns={'home_team':'friendly_matches_together'}).sort_values('friendly_matches_together',ascending=False).reset_index()
friendly_matches[:10]

home_team            away_team  friendly_matches_together
0      Hungary              Austria                         58
1      Austria              Hungary                         57
2  Netherlands              Belgium                         55
3      Belgium          Netherlands                         54
4        Kenya               Uganda                         46
5  El Salvador            Guatemala                         35
6      Belgium               France                         31
7   Martinique           Guadeloupe                         29
8       France              Belgium                         29
9       Guyana  Trinidad and Tobago                         28

**Austria** and **Hungary** have played most friendly matches together ***115*** matches followed by **Netherlands** and **Belgium** ***109*** matches.

## Goalscorers dataframe

In [25]:
goalscorers = pd.read_csv('/kaggle/input/international-football-results-from-1872-to-2017/goalscorers.csv')
goalscorers.sample(5)

date    home_team  away_team         team            scorer  \
30232  2011-10-07   Azerbaijan    Austria   Azerbaijan     Vüqar Nadirov   
18456  2000-02-19     Honduras       Peru     Honduras  José Luis Pineda   
32458  2013-10-11     Colombia      Chile        Chile    Alexis Sánchez   
6429   1974-06-26  Netherlands  Argentina  Netherlands      Johan Cruyff   
1019   1945-01-24        Chile    Bolivia        Chile    Juan Alcántara   

       minute  own_goal  penalty  
30232    74.0     False    False  
18456    69.0     False    False  
32458    29.0     False    False  
6429     11.0     False    False  
1019     75.0     False    False

In [26]:
goalscorers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41113 entries, 0 to 41112
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       41113 non-null  object 
 1   home_team  41113 non-null  object 
 2   away_team  41113 non-null  object 
 3   team       41113 non-null  object 
 4   scorer     41064 non-null  object 
 5   minute     40855 non-null  float64
 6   own_goal   41113 non-null  bool   
 7   penalty    41113 non-null  bool   
dtypes: bool(2), float64(1), object(5)
memory usage: 2.0+ MB


In [27]:
most_goal_players = pd.DataFrame(goalscorers.groupby(['team','scorer'])[['scorer','minute']].agg({'scorer':'count',
                                                                     'minute':'mean'})).rename(columns={'scorer':'goals_scored' , 'minute':'avg_minute'}).sort_values('goals_scored',ascending=False).reset_index()[:10]

most_goal_players

team              scorer  goals_scored  avg_minute
0             Portugal   Cristiano Ronaldo            91   53.857143
1               Poland  Robert Lewandowski            56   50.678571
2            Argentina        Lionel Messi            54   45.833333
3                 Iran            Ali Daei            49   56.653061
4              Germany      Miroslav Klose            48   48.250000
5            Guatemala         Carlos Ruiz            47   45.000000
6              Uruguay         Luis Suárez            46   52.478261
7               Sweden  Zlatan Ibrahimović            44   53.727273
8              England          Harry Kane            44   50.636364
9  Republic of Ireland        Robbie Keane            44   43.977273

**Cristiano Ronaldo** scored most goals(***91***) and has a average goal scoring time of 54 min(or mostly scores in 2nd half of the game). He is followed by **Robert Lewandowski** (***56 goals***) and average time 51 min(mostly scores in 2nd half) and **Lionel Messi**(***54 goals***) and average time 46 min(or we can say mostly scores at the end of 1st half)

### 10) Which player has scored more goals in penalty ?

In [28]:
penalty_goals = goalscorers[goalscorers['penalty'] == True]
penalty_goals.groupby(['team','scorer'])[['scorer']].agg({'scorer':'count',
                                                         }).rename(columns={'scorer':'goals_scored' }).sort_values('goals_scored',ascending=False).reset_index()[:10]

team              scorer  goals_scored
0             Portugal   Cristiano Ronaldo            16
1            Argentina        Lionel Messi            14
2             Bulgaria    Hristo Stoichkov            13
3              England          Harry Kane            13
4               Mexico   Cuauhtémoc Blanco            11
5               Poland  Robert Lewandowski            11
6            Australia        Mile Jedinak            11
7                Spain     Fernando Hierro            10
8  Republic of Ireland        Robbie Keane            10
9        United States      Landon Donovan            10

Again 1st spot is taken by **Cristiano Ronaldo** scored ***16*** goals followed by **Lionel Messi** - ***14*** goals and **Hristo Stoichkov** - ***13*** goals.

In [29]:
# parsing date 
goalscorers['date'] = pd.to_datetime(goalscorers['date'],errors='coerce',format='%Y-%m-%d')
goalscorers['Year'] = goalscorers['date'].dt.year

In [30]:
# which teams scored their own goals
own_goals = goalscorers[goalscorers['own_goal'] == True]
own_goal_teams = own_goals.groupby(['Year','team'])[['own_goal']].agg('count').sort_values('own_goal',ascending=False).rename(columns={'own_goal':'times_own_goal_scored'}).reset_index()

fig = px.line(own_goal_teams, x='Year', y=['times_own_goal_scored'],
              color='team',hover_name='team',
              markers = True,
              title="Teams that scored own goals through the years")

fig.update_xaxes(tickvals=np.arange(1917,2023,5),tickangle=45, tickfont=dict(family='Rockwell', size=14))
fig.show()

**Germany** has most own goals(4) followed by **Russia,Crotia,Mexico,Spain,Canada**.  

### 11) Which team or player score most goals in 1st half or second half ?

In [31]:
# dividing minute into 1st and 2nd half 
goalscorers['half'] = goalscorers['minute'].apply(lambda x: 'first_half' if (x < 45.0) else 'second_half')
goalscorers

date      home_team away_team           team            scorer  \
0     1916-07-02          Chile   Uruguay        Uruguay   José Piendibene   
1     1916-07-02          Chile   Uruguay        Uruguay  Isabelino Gradín   
2     1916-07-02          Chile   Uruguay        Uruguay  Isabelino Gradín   
3     1916-07-02          Chile   Uruguay        Uruguay   José Piendibene   
4     1916-07-06      Argentina     Chile      Argentina     Alberto Ohaco   
...          ...            ...       ...            ...               ...   
41108 2023-07-12        Jamaica    Mexico         Mexico       Luis Chávez   
41109 2023-07-12        Jamaica    Mexico         Mexico  Roberto Alvarado   
41110 2023-07-12  United States    Panama         Panama     Iván Anderson   
41111 2023-07-12  United States    Panama  United States    Jesús Ferreira   
41112 2023-07-16         Mexico    Panama         Mexico  Santiago Giménez   

       minute  own_goal  penalty  Year         half  
0        44.0     False    False  1916   first_half  
1        55.0     False    False  1916  second_half  
2        70.0     False    False  1916  second_half  
3        75.0     False    False  1916  second_half  
4         2.0     False    False  1916   first_half  
...       ...       ...      ...   ...          ...  
41108    30.0     False    False  2023   first_half  
41109    90.0     False    False  2023  second_half  
41110    99.0     False    False  2023  second_half  
41111   105.0     False    False  2023  second_half  
41112    88.0     False    False  2023  second_half  

[41113 rows x 10 columns]

In [32]:
team_goal_in_halfs = goalscorers.groupby(['team','half'])[['scorer']].agg('count').sort_values('scorer',ascending=False).reset_index().rename(columns={'scorer':'goals_scored'})

fig = px.bar(team_goal_in_halfs[:50],
             x='team', 
             y='goals_scored',
             color="half",
             title="Most goals scored in first and second halfs by teams(starting from highest)",
             text_auto=True)
fig.update_traces(textfont_size=12, textangle=0, textposition="outside", cliponaxis=False)
fig.show()

Top 5 spots shows that **Brazil,Germany,Argentina,Spain,Netherlands** had scored most goals in **2nd half** of the match because there are often player replacement or change of playing technique or goal chasing after the 1st half. 

In [33]:
player_goal_scored_in_halfs = pd.DataFrame(goalscorers.groupby(['scorer','half'])[['scorer']].agg('count')).rename(columns={'scorer':'goal_scored'}).sort_values('goal_scored',ascending=False).reset_index()

fig = px.bar(player_goal_scored_in_halfs[:50],
             x='scorer', 
             y='goal_scored',
             color="half",
             title="Most goals scored in first and second halfs by individual players(starting from highest)",
             text_auto=True)
fig.update_traces(textfont_size=12, textangle=0, textposition="outside", cliponaxis=False)
fig.show()

**Cristanio Ronaldo** is on 1st place and he mostly goals in second half.He is followed by **Ali Daei** and **Robert Lewandowski**. A point should be noted that these goals also include the penalty goals.